core/src/main/antlr4/com/databricks/labs/remorph/parsers/snowflake/SnowflakeParser.g4

```diff
@@ -148,6 +155,9 @@ otherCommand
     | let
     ;
 
+snowSqlCommand: BANG batch
+    ;
```

Comment:
This will parse `! <any valid sql batch>`, which includes an entire set of SQL statements, but does not include many of the actual commands listed in the Snowflake docs at: https://docs.snowflake.com/en/user-guide/snowsql-use

Here is an explanation of why I think we cannot implement this in the ANTLR grammar:

Looking through the actual available commands, they are all pretty standard simple command syntax, so you might first think, given that almost all keywords are allowed as identifiers, what you actually want is akin to:

```antlr
snowSqlCommand: BANG id exprList ;
```

But because some commands accept things that are not actually an expression, you need an expansion on expression for commands like this:

```
user#> !abort 77589bd1-bcbf-4ec8-9ebc-6c949b00614d;
```

Because the UUID is not a part of the standard syntax and we cannot make a token for this as it would eat 4-5. Here you only want to consume it without a syntax error, so specify the sequence and because the ! acts as a gate, it will not affect the rest of the grammar. So you will get to something like this:

```antlr
snowSqlCommand: BANG commandExpr (COMMA? commandExpr)* ;

commandExpr
    : (id | INTEGER)+ DOT  (id | INTEGER)+ DOT  (id | INTEGER)+ DOT  (id | INTEGER)+  // UUID 
    | expr
    ;
```

But this sequence is of indeterminate length as it relies on a newline to terminate it and newlines are not passed thorough to the Snowflake parser. So as written above, it would eat anything that looks like an expression, even on the next line as we humans see it (the parser just sees a sequence of tokens). The list is of indeterminate length because of `!print` and `!system`.  

Next, can we use a simple pre-processor? I think so. Convert it to `SQLCOMMAND "!system df;lkfglkjfklgj kl. "`

The BANG operator isn't a thing in Snowflake SQL, it is always NOT. However, It would be better avoid that. So I think that we need to use either lexer modes for this, or exact matches on tokens.

The problem  is that Snowflake already uses ! as a special character in system stored procedure calls, so if there is a clash there, then a simple lexer rule would not work. You can investigate that CALL:

```antlr
// Snowfllake allows calls to special internal stored procedures, named x.y!entrypoint
call: CALL objectName (BANG id)? L_PAREN exprList? R_PAREN
    ;
```
To see if there are any command clashes. If not then you can do this in the LEXER:

```antlr
SQLCOMMAND: '!' WS* ('abort' | 'system' | etc....)  ~[\r\n]*;
```

Note that you need to spell out all the tokens as otherwise the rule would eat the CALL special !name.

Add a test for each command and make sure to add a test that ensures CALL x.y!FRED() works.

If there are any clashes, then it is probably best if I take a deeper look but: `WS* ~'('` could be added before `~[\n\r]`

Reply:
implemented accordingly